### DISCUSSION TITLE
Support for Markdoc

### DISCUSSION THREAD

**Discussion by louiss0** [vuejs/vitepress#3135] [Ideas]
# Support for Markdoc

I'm here to ask for support for Markdoc.  Why Because Markdoc is a great tool that makes working with components a breeze. It's a Markdoc renderer that allows its user to configure what is and is not allowed in Markdown. It's highly performant and has a good Language Server. If you want to learn about it you can go to their [site](https://markdoc.dev/)  But I will highlight some features down below. 

## Tags

A tag is a special syntax that allows its user to write HTML or Components in a different form. They work just like any component. 

Self-closing tag.

```md
{% abbr "PHP" title="Hypertext Preprocessor" /%}
```

Paired tag
```md
{% details  %}

This is a set of details
{% /details %}
```
## Global Functions

You can configure global functions to be called in your Markdown. A global function works like this. 

```md
{debug("name")}
```

## Language Server

Markdoc's Language Server is good at catching bugs and knows about all the tags that you can use in Markdown. You can get auto complete for all the tags that you use. 

**Example**, using Markdoc to write a proper `<dl>` tag with `<dt>` and `<dd>` as proper children.
![demonstration-of-markdoc-validation-using-dd-tag](https://github.com/vuejs/vitepress/assets/29028601/79849a75-75bc-46f6-a5ea-1a7f471291de)

**Comment by brc-dd** [comment]
It won't be supported. We rely on a specific flavor of markdown and don't intend to switch to something entirely different. Also, since you can already use vue in markdown, most of these differences are irrelevant. Regarding language server, you can use Volar and enable vitepress support in it. If you have any specific thing in your mind, that's supported in markdoc, but doesn't have any equivalent in vitepress, then please create a feature request for that.

**Reply by louiss0** [reply]
I don't think Volar can detect global components in Markdown. If I'm wrong Can you show me how to set it up?

**Reply by brc-dd** [reply]
Global components won't show auto-completions most-likely, unless you maintain a .d.ts file mapping those exports. Local components and any vue syntax should work fine with `"vue.server.vitePress.supportMdFile": true`

**Reply by louiss0** [reply]
I did this PR because I thought this was the only way to get support for autocomplete in Markdown! 
Why was this not documented? Is the autocomplete of built-in components. going to be supported in the future? I don't know how to setup autocomplete for Vue projects.

**Reply by louiss0** [reply]
I'm trying to put all the types in the Global Components but I don't see the types for vitepress components. 
What file am I supposed to put this setting in "vue.server.vitePress.supportMdFile": true.

**Reply by louiss0** [reply]
Where are the types for the components Vitepress ships with I can't find them.

**Comment by louiss0** [comment]
I forgot If you want to toy around with Markdoc please clone this [repo](https://github.com/louiss0/dummy-markdoc-for-vitepress).

To get a similar experience you can Install the language server and an Extension called Error Lens. 
Just go to the `first.md` file and press <kbd>Ctrl</kbd> + <kbd>Space</kbd>

**Comment by louiss0** [comment]
> Global components won't show auto-completions most-likely, unless you maintain a .d.ts file mapping those exports. Local components and any vue syntax should work fine with `"vue.server.vitePress.supportMdFile": true`
